### ISSUE TITLE
Orthogonalization Issue?

### ISSUE THREAD

**Issue by felixGer** [HelmholtzAI-Consultants-Munich/PySDDR#2]
From my understanding, there is a discrepancy between the orthogonalization as presented in the SDDR paper, and the present implementation. Indeed, the proof of Lemma 1  seems to require an orthogonalization according to all structured components (no matter if those explicitely enter the network trunk or not). The present implementation, however, only performs the orthogonalization according to the structured effects that explicitely enter the deep neural network trunk (i.e. if x1 is a structured covariate that also serves as input to the network trunk, then U_hat is orthogonalized  with respect to x1. If no structured covariate enters the network trunk, then no orthogonalization takes place ). 
Am I missing anything in the code? Or is there any  reason  for this behavior? If yes, sorry for bothering! Otherwise, I believe the fix only requires a slight adjustment of the forward function in the sddrnetwork.py file (see pull request)

**Comment by davidruegamer** [comment]
Hi @felixGer . The idea is in fact that orthogonalization is only applied to the features that are also in the structured part -- because (only) for those we need to assure identifiability. The `xor` in Figure 1 in the newer version of the paper ([v4](https://arxiv.org/abs/2002.05777)) tries to make this more clear. 
Does that make sense to you?

**Comment by felixGer** [comment]
Hi @davidruegamer,

thank you very much for the fast reply! I understand why orthogonalization of the latent features is performed with respect to covariates that enter the structured part. However, i'm not sure to understand why orthogonalization of the latent features only takes place with respect to covariates that enter both the unstructured and structured part. 

Maybe the following examples will make my thought clearer:

1. Assume x1, x2, x3 are covariates to model outcome y. x1 enters the structured part while x2 and x3 enter the unstructured part. Assume x1 = x2 + x3. As x1 does not enter explicitely the unstructured part, no orthogonalization takes place. However, the structured part (linear effect of x1) is not identifiable (?)

2. In general, as there is no control about the values in the latent features (these result from model training), how can we be sure that these do not have perfect correlation with a covariate of the structured part? For instance, similarly to the example in the 'advanced tutorial',  let's take a covariate x1 being a number between 0-9, and an input to the unstructured part being a picture of the handwritten number in x1. As there are no shared inputs to the structured and unstructured part (at least explicitely), no orthogonalization takes place. Assume y = x1 + white noise. We model y = x1 + dnn(picture). If the dnn 'performs well', it might  pick up the entire information present in x1, in one of its latent features u, so that x1 is proportional to u. Making the linear effect of x1 on y unidentifiable (?)

**Comment by davidruegamer** [comment]
Good point @felixGer and definitely worth discussing in more detail! My general take on these kinds of problems first, before getting back to the original issue: 

- In the paper, we addressed identifiability issues that arise due to what you could call "redundancy in the model", i.e., both the structured and unstructured part (implicitly) include say a linear function of x1. 
- Unfortunately, identifiability problems that arise through predictor correlations ("redundancy in the features") are different and something that neither SDDR nor any other regression model can solve (without further information), as there is just no meaningful way to tell the different effects apart. Meaning that the identifiability issue of your example 1 would also exist for a simple regression model y ~ x1 + x2 + x3 (without any unstructured part) and it is then **up to the user** to figure out why s/he maybe runs into multicollinearity issues with this model and what covariates to exclude from/include into the model. NB: The good thing is that in the case of a "mild" correlation, the simple regression will be able to tell the effects apart (e.g., if you simulate x1 and x2 with correlation of 0.6, the linear model will still give you the correct effects on average) and so will SDDR.

So the question now boils down to whether or not the software should also deal with the second "redundancy". I am _not_ completely against this and indeed a safe call in both of your examples would be to orthogonalize w.r.t. every structured predictor. But this usually comes at the cost of less prediction performance accuracy, because more orthogonalization means more model restrictions and I am not sure if this is always in the user's interest. In particular, you will not be able to learn the joint effect of the unstructured and structured part, but the SDDR model will always learn the structured effects first, and then the unstructured part on what is left from the first step (roughly speaking).  

In the R version of SDDR I chose somewhat of a middle ground to deal with such cases but at the same time avoid being overly restrictive: the R package follows the same logic as currently implemented in PySDDR, but then the package additionally allows for custom orthogonalization so that user can, e.g., opt for the safe call "orthogonalize wrt. everything" option or maybe even an orthogonalization that s/he thinks makes most sense (e.g., explicitly orthogonalize input images wrt. to x2 and x7, because these might be correlated). 

Does this make sense to you?

**Comment by felixGer** [comment]
Yes, perfectly! Thanks for the clear answer.
